lib/commands/network.js

```diff
@@ -31,25 +35,48 @@ commands.isWifiOn = async function () {
 commands.setNetworkConnection = async function (type) {
   log.info("Setting network connection");
   // decode the input
-  let airplaneMode = type % 2;
-  type >>= 1;
-  let wifi = type % 2;
-  type >>= 1;
-  let data = type % 2;
+  const shouldEnableAirplaneMode = (type & AIRPLANE_MODE_MASK) !== 0;
+  const shouldEnableWifi = (type & WIFI_MASK) !== 0;
+  const shouldEnableDataConnection = (type & DATA_MASK) !== 0;
 
-  await this.wrapBootstrapDisconnect(async () => {
-    await this.adb.setAirplaneMode(airplaneMode);
-  });
-  await this.wrapBootstrapDisconnect(async () => {
-    await this.adb.broadcastAirplaneMode(airplaneMode);
-  });
-  if (!airplaneMode) {
+  const currentState = await this.getNetworkConnection();
+  const isArirplaneModeEnabled = (currentState & AIRPLANE_MODE_MASK) !== 0;
+  const isWiFiEnabled = (currentState & WIFI_MASK) !== 0;
+  const isDataEnabled = (currentState & DATA_MASK) !== 0;
+
+  if (shouldEnableAirplaneMode !== isArirplaneModeEnabled) {
+    await this.wrapBootstrapDisconnect(async () => {
+      await this.adb.setAirplaneMode(shouldEnableAirplaneMode);
+    });
     await this.wrapBootstrapDisconnect(async () => {
-      await this.setWifiState(wifi);
-      await this.adb.setDataState(data, this.isEmulator());
+      await this.adb.broadcastAirplaneMode(shouldEnableAirplaneMode);
     });
+  } else {
+    log.info('Not changing airplane mode, since it already has the correct state');
   }
 
+  if (shouldEnableAirplaneMode) {
+    log.info('Not changing data connection and Wi-Fi states, because airplane mode is enabled');
```

Comment:
Wi-Fi can be enabled in airplane mode on real devices.

Reply:
I'd consider such possibility as android bug then :)

Reply:
It's not a bug it's a feature (:
<img src="https://user-images.githubusercontent.com/3530074/38622562-18a1753e-3dac-11e8-8232-0ecc7358170d.png" height="50%" width="50%">

Reply:
Ok, I'll change that condition. I don't fly much though :)